common/metrics/otel_metrics_handler.go

```diff
@@ -35,40 +36,51 @@ import (
 	"go.temporal.io/server/common/log/tag"
 )
 
-// MetricsHandler is an event.Handler for OpenTelemetry metrics.
-// Its Event method handles Metric events and ignores all others.
-type otelMetricsHandler struct {
-	l           log.Logger
-	tags        []Tag
-	provider    OpenTelemetryProvider
-	excludeTags excludeTags
-}
+type (
+	// otelMetricsHandler is a Handler for OpenTelemetry metrics.
+	otelMetricsHandler struct {
+		l           log.Logger
+		tags        []Tag
+		provider    OpenTelemetryProvider
+		excludeTags excludeTags
+		catalog     catalog
+	}
+)
 
 var _ Handler = (*otelMetricsHandler)(nil)
 
-func NewOtelMetricsHandler(l log.Logger, o OpenTelemetryProvider, cfg ClientConfig) *otelMetricsHandler {
+func NewOtelMetricsHandler(
+	l log.Logger,
+	o OpenTelemetryProvider,
+	cfg ClientConfig,
+) (*otelMetricsHandler, error) {
+	c, err := globalRegistry.buildCatalog()
```

Comment:
This must be called after all metrics definitions are created. We may need to add some comments to explain that.
Also, there could be cases where user may use metrics handler to emit metric (using string as name) directly without defining the metric before hand. Should make sure they can continue to do so (of cause they won't have help text).

Reply:
> We may need to add some comments to explain that.

Added a comment that you must register the metrics with New*Def before calling this constructor in order for the metadata to be available.

> Also, there could be cases where user may use metrics handler to emit metric (using string as name) directly without defining the metric before hand. Should make sure they can continue to do so (of cause they won't have help text).

I added a comment saying that it's ok to use the handler on metrics you haven't registered too.